flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java

```diff
@@ -0,0 +1,304 @@
+      Iterator<Long> snapshotIdIterator = snapshotIdIterable.iterator();
+      if (snapshotIdIterator.hasNext()) {
+        this.startingSnapshotId = snapshotIdIterator.next();
+      }
+    }
+
+    this.inProgress = Sets.newHashSet();
+    Iterable<RewriteFileGroup> inProgressIterable = inProgressState.get();
+    if (inProgressIterable != null) {
+      for (RewriteFileGroup group : inProgressIterable) {
+        inProgress.add(group);
+      }
+    }
+  }
+
+  @Override
+  public void snapshotState(StateSnapshotContext context) throws Exception {
+    super.snapshotState(context);
+
+    startingSnapshotIdState.clear();
+    if (startingSnapshotId != null) {
+      startingSnapshotIdState.add(startingSnapshotId);
+    }
+
+    inProgressState.clear();
+    for (RewriteFileGroup group : inProgress) {
+      inProgressState.add(group);
+    }
+  }
+
+  @Override
+  public void processElement(StreamRecord<DataFileRewriteExecutor.ExecutedGroup> streamRecord) {
+    DataFileRewriteExecutor.ExecutedGroup executedGroup = streamRecord.getValue();
+    try {
+      if (commitService == null) {
+        this.commitService = createCommitService(executedGroup, streamRecord.getTimestamp());
+        this.startingSnapshotId = executedGroup.snapshotId();
+      }
+
+      commitService.offer(executedGroup.group());
+      inProgress.add(executedGroup.group());
+    } catch (Exception e) {
+      LOG.info(
+          "Exception processing {} for table {} with {}[{}] at {}",
+          executedGroup,
+          tableName,
+          taskName,
+          taskIndex,
+          streamRecord.getTimestamp(),
+          e);
+      output.collect(TaskResultAggregator.ERROR_STREAM, new StreamRecord<>(e));
+      errorCounter.inc();
+    }
+  }
+
+  @Override
+  public void processWatermark(Watermark mark) throws Exception {
+    try {
+      if (commitService == null && !inProgress.isEmpty()) {
+        this.commitService = createCommitService(null, mark.getTimestamp());
+      }
+
+      if (commitService != null) {
+        commitService.close();
+      }
+
+      table.refresh();
+      LOG.info(
+          "Successfully completed data file compaction to {} for table {} with {}[{}] at {}",
+          table.currentSnapshot().snapshotId(),
+          tableName,
+          taskName,
+          taskIndex,
+          mark.getTimestamp());
+    } catch (Exception e) {
+      LOG.info(
+          "Exception closing commit service for table {} with {}[{}] at {}",
+          tableName,
+          taskName,
+          taskIndex,
+          mark.getTimestamp(),
+          e);
+      output.collect(TaskResultAggregator.ERROR_STREAM, new StreamRecord<>(e));
+      errorCounter.inc();
+    }
+
+    // Cleanup
+    commitService = null;
+    startingSnapshotId = null;
+    inProgress.clear();
+
+    super.processWatermark(mark);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (commitService != null) {
+      commitService.close();
```

Comment:
There is no way to guarantee the consistency between the `inProgress` list and the data which is really committed to the Iceberg table. There could be an "old savepoint" which we are restoring, or there could be some delay between storing the state and stopping the job.

If we try to update the table and replace already replaced files, then the commit will fail. So we will not corrupt the table even in these cases.

Given these constraints I would prefer committing everything which is already done.

Reply:
> If we try to update the table and replace already replaced files, then the commit will fail. 

so first commit (with new rewrite group offered) after restore would always fail? if yes, that also doesn't seem ideal

> There is no way to guarantee the consistency between the inProgress list and the data which is really committed to the Iceberg table

wondering if commit should only happen in `snapshotState` with partial commit always disabled?

Reply:
We can't commit only on checkpoints. We need commit on watermarks too, so upcoming tasks will not conflict with pending commits waiting for checkpoint.

If we accept the statement above, that means that we still have errors on recovery when a watermark arrives on the middle of a checkpoint. On recovery the groups will be compacted again, and we will try to commit them and fail.

So, I don't see an error free solution. This is why I prefer to keep the partial commit feature, and live with the errors on recovery.

Maybe we don't increase the error counters during the first checkpoint? This seems like a bit hacky to me, but might help when monitoring a job

Reply:
> We can't commit only on checkpoints. We need commit on watermarks too, so upcoming tasks will not conflict with pending commits waiting for checkpoint.

can we add some metadata into snapshot summary to indicate which trigger has been committed. Then during restore, the committer can check if any restored groups has been committed already. similar to the Flink Iceberg sink using checkpointId to check the commit point. this may require disabling of partial commit, which is probably not necessary with Flink streaming jobs.

Reply:
What we need to do: Basically we would need to check every group if it was committed or not. This would mean to keep an up-to-date cache for the already committed groupIds (timestamp + globalId). This also could become "corrupt" if the grouping algorithm or configuration changes.

What we gain: without this we try and validation would prevent duplicate commits. This would happen only during restarts where unaligned checkpointing is used.

I'm starting to get convinced to basically just remove the state from the committer...
When "close" is called on the `CommitService` then every pending group is committed anyways. If we lose some groups, then we lose them... not a big deal. On job restart or on the next compaction they will be regenerated anyways.

Seems like it is less error prone if we just forget about the pending commits and try to commit every group which we receive.

Reply:
sure, I am fine with removing the state from the committer

Reply:
Removed